File: soundify/Controllers/WelcomeViewController.m
//
//  WelcomeViewController.m
//  soundify
//
//  Created by Raul Max on 01/03/22.
//

#import "WelcomeViewController.h"
#import "UIViewExtension.h"

@interface WelcomeViewController ()

@end

@implementation WelcomeViewController

- (UIButton *)signInButton {
  UIButton *button = [UIButton new];
  button.backgroundColor = [UIColor systemRedColor];
  button.layer.borderWidth = 2;
  button.clipsToBounds = YES;
  
  [button setTitle:@"Sign In with Spotify" forState:UIControlStateNormal];
  [button setTitleColor:[UIColor systemBackgroundColor] forState:UIControlStateNormal];
  
  return button;
}

- (void)viewDidLoad {
  [super viewDidLoad];
  // Do any additional setup after loading the view.
  
  self.title = @"Soundify";
  self.view.backgroundColor = [UIColor systemGreenColor];
  
  [self.view addSubview:self.signInButton];
}

- (void)viewDidLayoutSubviews {
  [super viewDidLayoutSubviews];
  
  self.signInButton.frame = CGRectMake(20, self.view.frame.size.height - 50 - self.view.safeAreaInsets.bottom, self.view.frame.size.width, 50);
  self.signInButton.layer.cornerRadius = self.signInButton.frame.size.height / 2;
  
  [self.signInButton addTarget:self action:@selector(handlePress:) forControlEvents:UIControlEventTouchUpInside];
  
  printf("%f", self.view.frame.size.height);
}

- (void)handlePress: (UIButton *) sender {
  printf("Button pressed!");
}

@end

File: Soundify/Controllers/Other/WelcomeViewController.swift
//
//  WelcomeViewController.swift
//  Soundify
//
//  Created by Raul Max on 02/12/21.
//

import UIKit

class WelcomeViewController: UIViewController {
  
  private let signInButton: UIButton = {
    let button = UIButton()
    button.backgroundColor = .none
    button.layer.borderWidth = 2
    button.clipsToBounds = true
    button.setTitle("Sign In with Spotify".uppercased(), for: .normal)
    button.setTitleColor(.systemBackground, for: .normal)
    return button
  }()
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    title = "Soundify"
    view.backgroundColor = .systemGreen
    view.addSubview(signInButton)
    signInButton.addTarget(self, action: #selector(didTapSignIn), for: .touchUpInside)
  }
  
  override func viewDidLayoutSubviews() {
    super.viewDidLayoutSubviews()
    signInButton.frame = CGRect(x: 20, y: view.height - 50 - view.safeAreaInsets.bottom, width: view.width - 40, height: 50)
    signInButton.layer.cornerRadius = signInButton.frame.size.height / 2
  }
  
  @objc func didTapSignIn() {
    let authVc = AuthViewController()
    authVc.completionHandler = { [weak self] success in
      self?.handleSignIn(success: success)
    }
    authVc.navigationItem.largeTitleDisplayMode = .never
    navigationController?.pushViewController(authVc, animated: true)
  }
  
  private func handleSignIn(success: Bool) {
    // Login or yell at the user
    guard success else {
      let alert = UIAlertController(title: "Oops...", message: "Something went wrong when trying to signing in.", preferredStyle: .alert)
      alert.addAction(UIAlertAction(title: "Dismiss", style: .cancel, handler: nil))
      present(alert, animated: true)
      return
    }
    
    let mainAppTabBarVc = TabBarViewController()
    mainAppTabBarVc.modalPresentationStyle = .fullScreen
    present(mainAppTabBarVc, animated: true)
  }
}
